### PR TITLE
feat: native eRPC integration via erpc-py

### PR DESCRIPTION
**Pull request by tars-endurance**
## Summary

Adds optional RPC caching proxy for Ursula nodes using [erpc-py](https://github.com/tars-endurance/erpc.py), a Python wrapper around [eRPC](https://erpc.cloud).

## What's included

- **`nucypher/utilities/rpc_proxy.py`** — Manages eRPC proxy lifecycle (start/stop) alongside the Ursula node
- **`tests/unit/test_rpc_proxy.py`** — Unit tests with mocked erpc-py dependency
- **`nucypher/constants.py`** — `NUCYPHER_ERPC_ENABLED` environment variable for feature flag
- **`newsfragments/43.feature.rst`** — Release note

## How it works

When `NUCYPHER_ERPC_ENABLED=true`, an eRPC caching proxy starts on a local port before the Ursula node connects to its RPC provider. The proxy transparently caches and deduplicates RPC requests, reducing load on upstream providers.

The feature is **opt-in** and disabled by default — no behavior change for existing deployments.

### CODE REVIEW

**Review by tars-endurance — COMMENTED** [review]
## 🔍 Code Review: eRPC Sidecar for Ursula Docker Container

Overall this is a well-designed, safety-conscious addition. The architecture is sound — memory-only cache, finality-aware TTLs, graceful bypass on failure, and transparent endpoint rewriting. The critical `TTL=0s` for unfinalized `eth_call` is correctly implemented. I have a few issues to address before merging.

---

### 🔴 Issues (Request Changes)

#### 1. Dockerfile: `COPY . /code` already includes `deploy/docker/` files — redundant COPY
Layer 5 copies `generate_erpc_config.py` and `entrypoint.sh` explicitly, but `COPY . /code` in Layer 3 already copies the entire repo including `deploy/docker/`. The explicit copies are redundant and could mask staleness issues if someone edits the file in the repo but forgets the separate COPY overwrites it. Either:
- Remove the redundant COPYs in Layer 5, or
- Move the explicit COPYs *before* `COPY . /code` for layer caching benefit and remove them from Layer 5

#### 2. Dockerfile: `RUN export PATH=...` is a no-op
`RUN export PATH="$HOME/.local/bin:$PATH"` sets PATH only for that RUN layer — it has no effect on subsequent layers or runtime. This was pre-existing but worth fixing now. Use `ENV PATH="$HOME/.local/bin:$PATH"` instead.

#### 3. Dockerfile: No integrity verification for eRPC binary
The eRPC binary is downloaded over HTTPS but without checksum verification. A supply-chain attack could inject a malicious binary. Add a SHA256 check:
```dockerfile
ARG ERPC_SHA256=<hash>
RUN sudo curl -L -o /usr/local/bin/erpc ... &&     echo "$ERPC_SHA256  /usr/local/bin/erpc" | sha256sum -c - &&     sudo chmod +x /usr/local/bin/erpc
```

#### 4. Entrypoint: `condition_blockchain_endpoints` not rewritten through eRPC
The entrypoint rewrites `ETH_ENDPOINT` and `POLYGON_ENDPOINT`, but if Ursula reads endpoints from its JSON config file (via `NUCYPHER_CONFIG_PATH`), those hardcoded URLs in the config won't be rewritten. Ursula would bypass eRPC for any endpoints read directly from the config file. This is the most significant correctness issue — the config generator reads the JSON to build eRPC's upstream list, but Ursula itself will still use the original URLs from that JSON config.

**Options:**
- Document that config-file users must set endpoints via env vars for eRPC to work, or
- Have the entrypoint rewrite the JSON config file (risky), or
- Have the entrypoint set `NUCYPHER_CONFIG_PATH` to empty and rely on env var fallback

#### 5. Shell script: unquoted `$GENERATE_ARGS` risks word splitting
`python3 ... $GENERATE_ARGS` will word-split. Use an array instead:
```bash
GENERATE_ARGS=(--output "$ERPC_CONFIG_PATH")
if [ -n "${NUCYPHER_CONFIG_PATH:-}" ] && [ -f "${NUCYPHER_CONFIG_PATH}" ]; then
    GENERATE_ARGS+=(--config "$NUCYPHER_CONFIG_PATH")
fi
python3 /code/deploy/docker/generate_erpc_config.py "${GENERATE_ARGS[@]}"
```

---

### 🟡 Suggestions

#### 6. Config generator: hardcoded chain ID assumptions
`eth_endpoint` is assumed to be chain 1 and `polygon_endpoint` chain 137. This is correct for mainnet but breaks for testnet deployments (Sepolia chain 11155111, Amoy chain 80002). The config generator should either:
- Accept chain ID overrides via env vars, or
- Document this limitation explicitly

#### 7. Tests: no test for `condition_blockchain_endpoints` with string values
`extract_endpoints` handles both `list` and `str` values for condition endpoints, but only the list case is tested. Add a test for the string case.

#### 8. Tests: no test for the YAML generation with unknown chains
`build_network_config` has an unknown chain fallback path tested, but `generate_erpc_yaml` with an unknown chain ID (e.g., Arbitrum 42161) isn't tested end-to-end.

#### 9. erpc.yaml template may be dead code
The static `erpc.yaml` template has `${ETH_ENDPOINT}` / `${POLYGON_ENDPOINT}` env var references, but the entrypoint always generates a config via `generate_erpc_config.py` and writes to `/tmp/erpc.yaml`. The template is never used at runtime. Is it intended as documentation only? If so, note that in its header. If it's a fallback, wire it up.

#### 10. Docker: eRPC binary adds ~30-50MB to image
Not blocking, but worth noting in the docs for operators who care about image size. The `NUCYPHER_ERPC_ENABLED=false` bypass doesn't reclaim the space.

---

### ✅ What's Good

- **TTL=0s for unfinalized eth_call** — exactly right for DKG safety
- **Graceful bypass** — every failure path falls through to original behavior
- **Signal handling** — proper trap + cleanup with TERM then KILL escalation
- **Health check with timeout** — correct pattern for sidecar readiness
- **Memory-only cache** — no external dependencies, bounded size
- **Tests are meaningful** — CLI integration tests, edge cases, validation
- **Documentation is clear** — covers all env vars, bypass mode, monitoring
- **No secrets in code** — endpoints from env vars only

**Review by KPrasch — DISMISSED** [review]
Personally, I find some of these files to be a little bit dizzy, particularly like unconventional Python doesn't follow industry best practices or conventions you know there's just a lot of wasted white space and I think we could make another pass at this. That is much more tidy and split the load between human concoction and machine readability.

**Review by KPrasch — DISMISSED** [review]
Personally, I find some of these files to be a little bit dizzy, particularly like unconventional Python doesn't follow industry best practices or conventions you know there's just a lot of wasted white space and I think we could make another pass at this. That is much more tidy and split the load between human concoction and machine readability. @tars-endurance

File: deploy/docker/entrypoint.sh [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+if [ -n "${NUCYPHER_CONFIG_PATH:-}" ] && [ -f "${NUCYPHER_CONFIG_PATH}" ]; then
+    GENERATE_ARGS="$GENERATE_ARGS --config $NUCYPHER_CONFIG_PATH"
+fi
+if [ -n "${ERPC_MAX_CACHE_ITEMS:-}" ]; then
+    GENERATE_ARGS="$GENERATE_ARGS --max-cache-items $ERPC_MAX_CACHE_ITEMS"
+fi
+
+if ! python3 /code/deploy/docker/generate_erpc_config.py $GENERATE_ARGS; then
+    log_warn "Failed to generate eRPC config — bypassing eRPC"
+    exec "$@"
```
[the remark this answers is not in the document]
Comment (tars-endurance) [reply deploy/docker/entrypoint.sh]: Fixed — switched to a bash array `generate_args=()` with proper quoting throughout.

File: tests/unit/test_rpc_proxy.py [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+        assert not proxy.is_active
+
+    @patch("nucypher.utilities.rpc_proxy.RPCProxy.start")
+    def test_fallback_on_import_error(self, mock_ursula_config):
+        """If erpc-py is not installed, start returns False."""
+        proxy = RPCProxy.from_ursula_config(mock_ursula_config)
+
+        with patch.dict("sys.modules", {"erpc": None}):
+            proxy.start()
+            # Start was mocked, so just verify proxy state
```
[the remark this answers is not in the document]
Comment (tars-endurance) [reply tests/unit/test_rpc_proxy.py]: Fixed in 1fe712f — replaced `@patch("RPCProxy.start")` with `sys.modules` stub pattern. The test now actually exercises the `ImportError` path in `start()` by setting `sys.modules["erpc"] = None`, which triggers the real fallback logic.

File: nucypher/utilities/rpc_proxy.py [hunk 3]
```diff
@@ -0,0 +1,353 @@
...
+            ``polygon_endpoint``, ``condition_blockchain_endpoints``, and ``domain``).
+        """
+        endpoints = collect_endpoints(
+            eth_endpoint=config.eth_endpoint,
+            polygon_endpoint=config.polygon_endpoint,
+            condition_blockchain_endpoints=config.condition_blockchain_endpoints,
+            domain=config.domain,
+        )
+
+        erpc_config = build_erpc_config(endpoints=endpoints)
```
[the remark this answers is not in the document]
Comment (tars-endurance) [reply nucypher/utilities/rpc_proxy.py]: Fixed — `from_ursula_config()` now wraps the `build_erpc_config()` call in a try/except ImportError, so callers get a clear error if erpc-py is missing rather than an unexpected crash. The `start()` method already had its own ImportError guard.

File: nucypher/utilities/rpc_proxy.py [hunk 4]
```diff
@@ -0,0 +1,353 @@
...
+    eth_endpoint: Optional[str],
+    polygon_endpoint: Optional[str],
+    condition_blockchain_endpoints: Optional[Dict[int, List[str]]],
+    domain,
+) -> Dict[int, List[str]]:
+    """Collect all chain endpoints from Ursula's configuration into a unified map.
+
+    Mirrors the logic in ``UrsulaConfiguration.configure_condition_blockchain_endpoints``
+    without modifying any config state.
+
```
[the remark this answers is not in the document]
Comment (tars-endurance) [reply nucypher/utilities/rpc_proxy.py]: Fixed — changed `insert(0, ...)` to `append(...)` to match the ordering behavior in `UrsulaConfiguration.configure_condition_blockchain_endpoints`.

File: tests/unit/test_rpc_proxy.py [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+
+    @patch("nucypher.utilities.rpc_proxy.build_erpc_config.__module__", "nucypher.utilities.rpc_proxy")
+    def test_build_config(self, sample_endpoints):
+        """Verify config is built with correct parameters."""
+        config = build_erpc_config(endpoints=sample_endpoints)
+        # These come from erpc.ERPCConfig
+        assert config.project_id == "taco-ursula"
+        assert config.server_port == 4000
+        assert config.cache.method_ttls["eth_call"] == 0  # DKG safety
+        assert config.cache.method_ttls["eth_getLogs"] == 2
```
[the remark this answers is not in the document]
Comment (tars-endurance) [reply tests/unit/test_rpc_proxy.py]: Fixed — all tests that call `build_erpc_config()` now use `patch.dict(sys.modules, {"erpc": _erpc_stub})` so they work regardless of whether erpc-py is installed. The stub is a module-level `MagicMock`.

**Review by KPrasch — COMMENTED** [review]
@tars-endurance I see that you have a lot of heading style comments inside of the RPC proxy maybe that's part of your code style, but I'm wondering if what that really indicates is that you need more space? While it's elegant that it's one file  - does it really want to be more than one file?

File: deploy/docker/generate_erpc_config.py [hunk 6]
```diff
@@ -0,0 +1,347 @@
...
+    lines.append("        ttl: 1m")
+    lines.append("")
+    lines.append("server:")
+    lines.append('  httpHostV4: "0.0.0.0"')
+    lines.append("  httpPortV4: 4000")
+    lines.append("  maxTimeout: 60s")
+    lines.append("")
+    lines.append("metrics:")
+    lines.append("  enabled: true")
+    lines.append('  hostV4: "0.0.0.0"')
```
[the remark this answers is not in the document]
Comment (tars-endurance) [reply deploy/docker/generate_erpc_config.py]: Already using `127.0.0.1` — the erpc.yaml / generate_erpc_config.py files were removed in favor of programmatic config via erpc-py.

File: nucypher/cli/commands/ursula.py [hunk 7]
```diff
@@ -649,10 +649,20 @@ def run(
             collection_interval=metrics_interval,
         )
 
+    # eRPC RPC Proxy — infrastructure dependency, must start before
+    # any blockchain connections are created during character init.
+    _rpc_proxy = None
+    from nucypher.utilities.rpc_proxy import is_erpc_enabled
+    if is_erpc_enabled():
+        _rpc_proxy = _start_rpc_proxy(emitter, character_options, config_file)
```
[the remark this answers is not in the document]
Comment (tars-endurance) [reply nucypher/cli/commands/ursula.py]: Fixed — moved `BlockchainInterfaceFactory` import to module level (line 72). The function now uses the module-scope import cleanly.
